### PR TITLE
Slight tweaks in handling of OperationCanceledExceptions during analy… 

### PR DESCRIPTION
**Pull request by mavasani**
…zer execution:

1. Don't swallow OperationCanceledException fired with our cancellelationToken. These are supposed to be handled up the stack and swallowing them would mean unnecessary computation would be done for result that would eventually be ignored.
2. Analyzer callbacks that register actions need custom handling of all OperationCanceledExceptions in AnalyzerManager. Multiple threads trying to fetch registered analyzer actions might share the same task for computing registered actions, and if cancellation is fired on the source of this task, we want to re-attempt invoking the register callbacks, otherwise the analyzer will never execute.

@JohnHamby @pharring @srivatsn @tmeschter @heejaechang can you please review?

### CODE REVIEW

File: src/Compilers/Core/AnalyzerDriver/AnalyzerExecutor.cs [hunk 1]
```diff
@@ -64,20 +65,41 @@ internal class AnalyzerExecutor
...
                 addDiagnostic: null,
                 onAnalyzerException: onAnalyzerException,
+                calleeHandlesOperationCanceledException: false,
                 cancellationToken: cancellationToken);
         }
 
+        /// <summary>
+        /// Returns a cloned instance of <see cref="AnalyzerExecutor"/>, except the fact that it doesn't handle <see cref="OperationCanceledException"/> for callbacks into the executor.
+        /// Callee of the AnalyzerExecutor methods should handle <see cref="OperationCanceledException"/>.
+        /// </summary>
```
Comment (heejaechang) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerExecutor.cs]: can we just use AsyncLazy here?

**Comment by heejaechang** [comment]
I would just make AsyncLazy down to shared layer and use it.

**Comment by mavasani** [comment]
Moving to AsyncLazy is tracked by #664 . However, for this case even AsyncLazy doesn't help if the core AnalyzerExecutor.ExecuteAndCatchIfThrows swallows the OperationCanceledException as the task will never fault. We want OperationCanceledException to bubble up when invoking register methods so the caller can handle it and re-attempt invoking these methods.

**Comment by heejaechang** [comment]
@mavasani sure, (1) must be done. cancellation token should never be swallowed. but (2), I think it should use AsyncLazy.

**Comment by mavasani** [comment]
@heejaechang Actually I have put some more thought into this, and I think (2) is not even a problem that current exists. The only part wrapped within a try-catch is the core analyzer callback to register actions, and that should not throw an OperationCanceledOperation with a cancellation token different then one passed into the context argument passed to the analyzer. If it does, then the analyzer is doing something wrong and we ought to catch and report a diagnostic for that exception. Outer task invoking the register methods in AnalyzerManager might fault with OperationCanceledException with a differing cancellation token, but that won't be caught by AnalyzerExecutor and we already catch and handle that in AnalyzerManager. I think we only need to do (1) here, I'll send a new iteration.

**Comment by mavasani** [comment]
And yes, we will move this code to use AsyncLazy, but after we snap for rc2 so it has enough bake time.

**Comment by mavasani** [comment]
I have updated the pull request so we now only do (1).

**Comment by heejaechang** [comment]
:+1:

**Comment by pharring** [comment]
Nice. :+1:

**Comment by srivatsn** [comment]
:+1:

**Comment by tmeschter** [comment]
:+1:

**Comment by JohnHamby** [comment]
Maybe it's too late to matter, but I sign off.

**Pull request merged**